### PR TITLE
Update documentation

### LINKED ISSUE #7: Update colab notebooks
Notebooks are here: https://github.com/Stable-Baselines-Team/rl-colab-notebooks

Todo:

- [x] getting started
- [x] saving/loading
- [x] multiprocessing
- [x] monitor training
- [x] atari games
- [x] pybullet notebook
- [x] hindsight experience replay (requires HER to be implemented)
- [x] rl zoo

New colab:

- [x] saving/loading policies + eval env

Tutorial:

- [x] getting started
- [x] gym wrappers
- [x] multiprocessing
- [x] callbacks
- [x] custom gym env

### LINKED ISSUE #166: Missing Documentation
See #1

- [x] Custom Policies (complete example)
- [x] Migration guide (#123 )
- [x] get/set parameters examples (see #138 )
- [x] create eval env
- [x] save/loading of policy alone
- [x] save/load replay buffer
- [x] Clean up type hints from docstrings (see #167) 
- [x]  Fix autodoc for policies (cf TQC sb3 contrib)
- [x]  Add results section to algorithms

### PR DESCRIPTION
**Pull request by araffin**
## Description

closes #7 
closes #166 

Missing:
- [x] get/set parameters examples (see #138 )
- [x] save/load replay buffer
- [x]  Add results section to algorithms

NOTE: in the get/set parameters, I used PyTorch native methods, as it was easier and cleaner.

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: docs/guide/examples.rst [hunk 1]
```diff
@@ -491,6 +491,98 @@ Behind the scene, SB3 uses an :ref:`EvalCallback <callbacks>`.
...
+      learning_rate=0.05,
+  )
+
+  # Use traditional actor-critic policy gradient updates to
+  # find good initial parameters
+  model.learn(total_timesteps=10000)
+
+  # Include only variables with "policy", "action" (policy) or "shared_net" (shared layers)
+  # in their name: only these ones affect the action.
+  # NOTE: you can retrieve those parameters using model.get_parameters() too
```
Comment (Miffyli) [inline docs/guide/examples.rst:545]: Hmm we end up with many options for modifying parameters, should probably differentiate what is the right way of doing it / what are the differences. Maybe this example could use `get_parameters` as the goal is to contain (almost) all necessary parameters for value predictions etc.

Comment (araffin) [reply docs/guide/examples.rst:545]: That's documented in the export section ;)
I would say that PyTorch internal method is recommended in most cases but `get_parameters()` is needed if you want everything (including optimizers).

>Maybe this example could use get_parameters as the goal is to contain (almost) all necessary parameters for value predictions etc.

in practice, it is not so clean with `get_parameters()`...

Comment (Miffyli) [reply docs/guide/examples.rst:545]: Fair enough :). I agree torch's way is cleaner (and especially with the neat vectorization functions for policy), which we should promote first whenever possible.

File: docs/guide/custom_policy.rst [hunk 2]
```diff
@@ -7,6 +7,12 @@ Stable Baselines3 provides policy networks for images (CnnPolicies)
 and other type of input features (MlpPolicies).
 
 
+.. warning::
+  For all algorithms (except DDPG, TD3 and SAC), continuous actions are clipped during training and testing
+  (to avoid out of bound error).
```
Comment (Miffyli) [inline docs/guide/custom_policy.rst:12]: Could mention what DDPG/TD3/SAC do differently (squashing?).

File: docs/guide/examples.rst [hunk 3]
```diff
@@ -417,6 +419,170 @@ The parking env is a goal-conditioned continuous control task, in which the vehi
...
+
+By default, the replay buffer is not saved when calling ``model.save()``, in order to save space on the disk (a replay buffer can be up to several GB when using images).
+However, SB3 provides a ``save_replay_buffer()`` and ``load_replay_buffer()`` method to save it separately.
+
+
+.. image:: ../_static/img/colab-badge.svg
+   :target: https://colab.research.google.com/github/Stable-Baselines-Team/rl-colab-notebooks/blob/sb3/advanced_saving_loading.ipynb
+
+Stable-Baselines3 allows to automatically create an environment for evaluation.
+For that, you only need to specify ``create_eval_env=True`` when passing the Gym ID of the environment.
```
Comment (Miffyli) [inline docs/guide/examples.rst:437]: "passing the Gym ID of the environment" -> "passing the Gym ID of the environment while creating the agent."

File: docs/guide/examples.rst [hunk 4]
```diff
@@ -417,6 +419,170 @@ The parking env is a goal-conditioned continuous control task, in which the vehi
...
+  model.save_replay_buffer("sac_replay_buffer")
+
+  # load it into the loaded_model
+  loaded_model.load_replay_buffer("sac_replay_buffer")
+
+  # now the loaded replay is not empty anymore
+  print(f"The loaded_model has {loaded_model.replay_buffer.size()} transitions in its buffer")
+
+  # Save the policy independently from the model
+  # Note: if you don't save the complete model, you cannot continue training afterward
```
Comment (Miffyli) [inline docs/guide/examples.rst:472]: "if you don't save the complete model" -> "if you don't save the complete model with `model.save(...)`

File: docs/guide/export.rst [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+Stable Baselines3 does not include tools to export models to other frameworks, but
+this document aims to cover parts that are required for exporting along with
+more detailed stories from users of Stable Baselines3.
+
+
+Background
+----------
+
+In Stable Baselines3, the controller is stored inside policies which convert
+observations into actions. Each learning algorithm (e.g. DQN, A2C, SAC) contains
```
Comment (Miffyli) [inline docs/guide/export.rst:18]: "Each learning algorithm (e.g. DQN, A2C, SAC) contains one policy". I would replace this with something that clarifies this is the main behaviour policy that decides acts when doing e.g. `model.predict`. E.g:

"Each learning algorithm (e.g. DQN, A2C, SAC) contains a policy which represents the currently learned behaviour, accessible via ``model.policy``. "

File: docs/guide/export.rst [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+one policy, accessible via ``model.policy``.
+
+Policies hold enough information to do the inference (i.e. predict actions),
+so it is enough to export these policies (cf :ref:`examples <examples>`)
+to do inference in an another framework.
+
+.. warning::
+  When using CNN policies, the observation is normalized during pre-preprocessing (dividing by 255 to have values in [0, 1])
```
Comment (Miffyli) [inline docs/guide/export.rst:26]: Clarification: The processing is done _inside_ policy. E.g: "When using CNN policies, the observation is normalized inside the policy code (dividing by 255 to have values in [0, 1])".

**Comment by Miffyli** [comment]
Left some minor comments but mostly looks good :). With this amount of text I would run modifications through grammar-check just to be on the safe-side.

**Comment by araffin** [comment]
>Left some minor comments but mostly looks good :). With this amount of text I would run modifications through grammar-check just to be on the safe-side.

Thanks for the comments =)
I ran the spell-check already... Do you have a good to recommend for grammar-check?

**Comment by Miffyli** [comment]
> I ran the spell-check already... Do you have a good to recommend for grammar-check?

Not really, I just personally copy-paste things to Word or GDocs :'D . Should probably invest in "proper" tools.

**Comment by Miffyli** [comment]
Ran text through GDocs for crude grammar check. From my side this is good to go, feel free to approve and merge if you agree with the changes :)
